File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/TimePeriodContainer.java
package kr.kth.commons.timeperiod;

import com.google.common.base.Function;
import com.google.common.collect.Lists;
import kr.kth.commons.base.Guard;
import kr.kth.commons.base.SortDirection;
import kr.kth.commons.timeperiod.tools.TimeTool;
import kr.kth.commons.tools.ListTool;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;

import java.util.*;

/**
 * {@link ITimePeriod}를 항목으로 가지는 컨테이너를 표현합니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 21.
 */
@Slf4j
public class TimePeriodContainer implements ITimePeriodContainer {

	private static final long serialVersionUID = -7661743810786491113L;

	protected final List<ITimePeriod> periods = Lists.newArrayList();

	protected TimePeriodContainer() {}

	protected TimePeriodContainer(Iterable<? extends ITimePeriod> periods) {
		this.periods.addAll(Lists.newArrayList(periods));
	}

	protected TimePeriodContainer(ITimePeriod... periods) {
		this.periods.addAll(Lists.newArrayList(periods));
	}


	@Override
	public void setStart(DateTime start) {
		if (size() > 0)
			move(start.getMillis() - getStart().getMillis());
	}

	@Override
	public void setEnd(DateTime end) {
		if (size() > 0)
			move(getEnd().getMillis() - end.getMillis());
	}

	@Override
	public boolean containsPeriod(ITimePeriod target) {
		return this.periods.contains(target);
	}

	@Override
	public void addAll(Iterable<? extends ITimePeriod> periods) {
		this.periods.addAll(Lists.newArrayList(periods));
	}

	@Override
	public void sortByStart(SortDirection direction) {
		if (direction == SortDirection.ASC) {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o1.getStart().compareTo(o2.getStart());
				}
			});
		} else {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o2.getStart().compareTo(o1.getStart());
				}
			});
		}
	}

	@Override
	public void sortByEnd(SortDirection direction) {
		if (direction == SortDirection.ASC) {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o1.getEnd().compareTo(o2.getEnd());
				}
			});
		} else {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o2.getStart().compareTo(o1.getEnd());
				}
			});
		}
	}

	@Override
	public void sortByDuration(SortDirection direction) {
		if (direction == SortDirection.ASC) {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o1.getDuration().compareTo(o2.getDuration());
				}
			});
		} else {
			Collections.sort(periods, new Comparator<ITimePeriod>() {
				@Override
				public int compare(ITimePeriod o1, ITimePeriod o2) {
					return o2.getDuration().compareTo(o1.getDuration());
				}
			});
		}
	}

	@Override
	public DateTime getStart() {
		if (size() > 0) {
			List<DateTime> starts =
				Lists.transform(periods,
				                new Function<ITimePeriod, DateTime>() {
					                @Override
					                public DateTime apply(ITimePeriod input) {
						                return input.getStart();
					                }
				                });
			return (DateTime) ListTool.min(starts);
		}

		return TimeSpec.MinPeriodTime;
	}

	@Override
	public DateTime getEnd() {
		if (size() > 0) {
			List<DateTime> ends =
				Lists.transform(periods,
				                new Function<ITimePeriod, DateTime>() {
					                @Override
					                public DateTime apply(ITimePeriod input) {
						                return input.getEnd();
					                }
				                });
			return (DateTime) ListTool.max(ends);
		}
		return TimeSpec.MaxPeriodTime;
	}

	@Override
	public Long getDuration() {
		return hasPeriod() ? (getEnd().getMillis() - getStart().getMillis())
		                   : TimeSpec.MaxPeriodDuration;
	}

	@Override
	public String getDurationDescription() {
		return TimeFormatter.getInstance().getDuration(getDuration(), DurationFormatKind.Detailed);
	}

	@Override
	public boolean hasStart() {
		return getStart() != TimeSpec.MinPeriodTime;
	}

	@Override
	public boolean hasEnd() {
		return getEnd() != TimeSpec.MaxPeriodTime;
	}

	@Override
	public boolean hasPeriod() {
		return hasStart() && hasEnd();
	}

	@Override
	public boolean isMoment() {
		return hasPeriod() && Objects.equals(getStart(), getEnd());
	}

	@Override
	public boolean isAnyTime() {
		return !hasStart() && !hasEnd();
	}

	@Override
	public boolean isReadonly() {
		return false;
	}

	@Override
	public void setup(DateTime start, DateTime end) {
		throw new IllegalStateException("TimePeriodContainer는 setup() 메소드를 지원하지 않습니다.");
	}

	@Override
	public ITimePeriod copy(long offset) {
		throw new IllegalStateException("TimePeriodContainer는 copy(long offset) 메소드를 지원하지 않습니다.");
	}

	@Override
	public void move(long offset) {
		if (offset == 0 || size() == 0) return;

		for (ITimePeriod period : periods)
			period.move(offset);
	}

	@Override
	public boolean isSamePeriod(ITimePeriod that) {
		return (that != null) &&
			       Objects.equals(getStart(), that.getStart()) &&
			       Objects.equals(getEnd(), that.getEnd());
	}

	@Override
	public boolean hasInside(DateTime moment) {
		return TimeTool.hasInside(this, moment);
	}

	@Override
	public boolean hasInside(ITimePeriod that) {
		return TimeTool.hasInside(this, that);
	}

	@Override
	public boolean intersectsWith(ITimePeriod that) {
		return TimeTool.intersectsWith(this, that);
	}

	@Override
	public boolean overlapsWith(ITimePeriod that) {
		return TimeTool.overlapsWith(this, that);
	}

	@Override
	public void reset() {
		this.periods.clear();
	}

	@Override
	public PeriodRelation getRelation(ITimePeriod that) {
		return TimeTool.getRelation(this, that);
	}

	@Override
	public String getDescription(ITimeFormatter formatter) {
		return format(formatter);
	}

	public String format(ITimeFormatter formatter) {
		if (formatter == null)
			formatter = TimeFormatter.getInstance();

		return formatter.getCollectionPeriod(size(), getStart(), getEnd(), getDuration());
	}

	@Override
	public ITimePeriod getIntersection(ITimePeriod that) {
		return TimeTool.getIntersectionRange(this, that);
	}

	@Override
	public ITimePeriod getUnion(ITimePeriod that) {
		return TimeTool.getUnionRange(this, that);
	}

	@Override
	public int compareTo(ITimePeriod that) {
		Guard.shouldNotBeNull(that, "that");
		return getStart().compareTo(that.getStart());
	}

	@Override
	public int size() {
		return this.periods.size();
	}

	@Override
	public boolean isEmpty() {
		return this.periods.isEmpty();
	}

	@Override
	public boolean contains(Object o) {
		return this.periods.contains(o);
	}

	@Override
	public Iterator<ITimePeriod> iterator() {
		return this.periods.iterator();
	}

	@Override
	public Object[] toArray() {
		return this.periods.toArray();
	}

	@Override
	public <T> T[] toArray(T[] a) {
		return this.periods.toArray(a);
	}

	@Override
	public boolean add(ITimePeriod period) {
		return period != null && this.periods.add(period);
	}

	@Override
	public boolean remove(Object o) {
		return o != null && this.periods.remove(o);
	}

	@Override
	public boolean containsAll(Collection<?> c) {
		return c != null && this.periods.containsAll(c);
	}

	@Override
	public boolean addAll(Collection<? extends ITimePeriod> c) {
		return c != null && this.periods.addAll(c);
	}

	@Override
	public boolean addAll(int index, Collection<? extends ITimePeriod> c) {
		return c != null && this.periods.addAll(index, c);
	}

	@Override
	public boolean removeAll(Collection<?> c) {
		return c != null && this.periods.removeAll(c);
	}

	@Override
	public boolean retainAll(Collection<?> c) {
		return c != null && this.periods.retainAll(c);
	}

	@Override
	public void clear() {
		this.periods.clear();
	}

	@Override
	public ITimePeriod get(int index) {
		return this.periods.get(index);
	}

	@Override
	public ITimePeriod set(int index, ITimePeriod element) {
		return this.periods.set(index, element);
	}

	@Override
	public void add(int index, ITimePeriod element) {
		this.periods.add(index, element);
	}

	@Override
	public ITimePeriod remove(int index) {
		return this.periods.remove(index);
	}

	@Override
	public int indexOf(Object o) {
		return this.periods.indexOf(o);
	}

	@Override
	public int lastIndexOf(Object o) {
		return this.periods.lastIndexOf(o);
	}

	@Override
	public ListIterator<ITimePeriod> listIterator() {
		return this.periods.listIterator();
	}

	@Override
	public ListIterator<ITimePeriod> listIterator(int index) {
		return this.periods.listIterator(index);
	}

	@Override
	public List<ITimePeriod> subList(int fromIndex, int toIndex) {
		return this.periods.subList(fromIndex, toIndex);
	}
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/TimePeriodContainer.scala
package kr.kth.timeperiod

import java.util
import kr.kth.commons.slf4j.Logging
import kr.kth.commons.tools.ScalaHash
import kr.kth.commons.{ValueObjectBase, Guard, SortDirection}
import org.joda.time.DateTime
import scala.collection.JavaConversions._
import scala.collection.{JavaConversions, TraversableOnce}

/**
 * kr.kth.timeperiod.TimePeriodContainer
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 14
 */
class TimePeriodContainer extends ITimePeriodContainer {

  def this(periods: Iterable[ITimePeriod]) {
    this()
    _periods ++= periods
  }
}

object TimePeriodContainer extends Logging {

  def apply(): TimePeriodContainer = new TimePeriodContainer()

  def apply(periods: Iterable[ITimePeriod]): TimePeriodContainer = new TimePeriodContainer(periods)

  def apply(periods: ITimePeriod*): TimePeriodContainer = apply(periods.toIterable)
}

/**
 * {@link ITimePeriod} 를 요소로 가지는 컨테이너를 나타냅니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 13.
 */
trait ITimePeriodContainer extends ValueObjectBase with scala.collection.Iterable[ITimePeriod] with ITimePeriod {

  protected var _periods = collection.mutable.ArrayBuffer[ITimePeriod]()

  protected def getPeriods: collection.mutable.ArrayBuffer[ITimePeriod] = _periods

  def iterator: Iterator[ITimePeriod] = _periods.iterator

  def reverseIterator: Iterator[ITimePeriod] = _periods.reverseIterator

  def add(elem: ITimePeriod) { append(elem) }

  def append(elems: ITimePeriod*) { appendAll(elems) }

  /**
   * 지정한 {@link ITimePeriod} 들을 컨테이너에 모두 추가한다.
   */
  def appendAll(items: TraversableOnce[ITimePeriod]) {
    _periods.appendAll(items)
  }

  def clear() { _periods.clear() }

  def contains(elem: ITimePeriod): Boolean = _periods.contains(elem)

  /**
   * 지정한 기간을 포함하고 있는지 검사한다.
   */
  def containsPeriod(target: ITimePeriod): Boolean = _periods.contains(target)


  def remove(elem: ITimePeriod*) { _periods --= elem; }

  // def remove(n: Int) { _periods.remove(n) }

  override def size = _periods.size

  override def isReadonly = false

  override def isEmpty: Boolean = (size == 0)

  def get(index: Int): ITimePeriod = _periods.get(index)

  def set(index: Int, element: ITimePeriod): ITimePeriod = {
    _periods.set(index, element)
    _periods(index)
  }

  def indexOf(o: Any): Int = if (o.isInstanceOf[ITimePeriod]) _periods.indexOf(o) else -1

  def lastIndexOf(o: Any): Int = if (o.isInstanceOf[ITimePeriod]) _periods.lastIndexOf(o) else -1

  def listIterator(): java.util.ListIterator[ITimePeriod] =
    JavaConversions.bufferAsJavaList(_periods).listIterator()

  def listIterator(index: Int): util.ListIterator[ITimePeriod] =
    JavaConversions.bufferAsJavaList(_periods).listIterator(index)

  def subList(fromIndex: Int, toIndex: Int): java.util.List[ITimePeriod] =
    _periods.subList(fromIndex, toIndex)

  /**
   * 모든 항목을 {@link ITimePeriod#getStart()} 를 기준으로 정렬합니다.
   */
  def sortByStart(direction: SortDirection = SortDirection.ASC) {
    direction match {
      case SortDirection.ASC =>
        _periods = _periods.sortWith(_ < _) // _.getStart.getMillis < _.getStart.getMillis)
      case SortDirection.DESC =>
        _periods = _periods.sortWith(_ > _) //_.getStart.getMillis > _.getStart.getMillis)
    }
  }

  /**
   * 모든 항목을 {@link ITimePeriod#getEnd()} 를 기준으로 정렬합니다.
   */
  def sortByEnd(direction: SortDirection = SortDirection.ASC) {
    direction match {
      case SortDirection.ASC =>
        _periods = _periods.sortWith(_.getEnd.getMillis < _.getEnd.getMillis)
      case SortDirection.DESC =>
        _periods = _periods.sortWith(_.getEnd.getMillis > _.getEnd.getMillis)
    }
  }

  /**
   * 모든 항목을 {@link ITimePeriod#getDuration()} 를 기준으로 정렬합니다.
   */
  def sortByDuration(direction: SortDirection = SortDirection.ASC) {
    direction match {
      case SortDirection.ASC =>
        _periods = _periods.sortWith(_.getDuration < _.getDuration)
      case SortDirection.DESC =>
        _periods = _periods.sortWith(_.getDuration > _.getDuration)
    }
  }

  override def getStart: DateTime =
    if (size > 0) _periods.minBy(x => x.getStart.getMillis).getStart
    else TimeSpec.MinPeriodTime

  def setStart(newStart: DateTime) {
    if (size > 0)
      move(newStart.getMillis - getStart.getMillis)
  }

  override def getEnd: DateTime =
    if (size > 0) _periods.maxBy(x => x.getEnd.getMillis).getEnd
    else TimeSpec.MaxPeriodTime

  def setEnd(newEnd: DateTime) {
    if (size > 0)
      move(newEnd.getMillis - getStart.getMillis)
  }

  override def getDuration: Long =
    if (hasPeriod) getEnd.getMillis - getStart.getMillis
    else TimeSpec.MaxPeriodDuration

  override def getDurationDescription =
    TimeFormatter.instance.getDuration(this.getDuration, DurationFormatKind.Detailed)

  override def hasStart = (size > 0) && (getStart != TimeSpec.MinPeriodTime)

  override def hasEnd = (size > 0) && (getEnd != TimeSpec.MaxPeriodTime)

  override def hasPeriod = (hasStart && hasEnd)

  override def isMoment = (hasPeriod) && (getStart == getEnd)

  override def isAnytime = (!hasStart && !hasEnd)

  override def setup(newStart: DateTime, newEnd: DateTime) {
    throw new UnsupportedOperationException("ITimePeriodContainer에서는 setup() 메소드를 지원하지 않습니다.")
  }

  override def copy(offset: Long) = {
    throw new UnsupportedOperationException("ITimePeriodContainer에서는 copy() 메소드를 지원하지 않습니다.")
  }

  override def move(offset: Long) {
    if (offset != 0L)
      _periods.foreach(_.move(offset))
  }

  def isSampePeriod(other: ITimePeriod): Boolean = {
    Guard.shouldNotBeNull(other, "other")
    (getStart == other.getStart) && (getEnd == other.getEnd)
  }

  override def hasInside(moment: DateTime): Boolean =
    Times.hasInside(this, moment)

  override def hasInside(other: ITimePeriod): Boolean =
    Times.hasInside(this, other)

  override def intersectsWith(other: ITimePeriod): Boolean =
    Times.intersectsWith(this, other)

  override def overlapsWith(other: ITimePeriod): Boolean =
    Times.overlapsWith(this, other)

  override def reset() {
    _periods.clear()
  }

  override def getRelation(other: ITimePeriod) =
    Times.getRelation(this, other)

  override def getDescription(formatter: Option[ITimeFormatter] = None) =
    format(formatter)

  override protected def format(formatter: Option[ITimeFormatter]) =
    formatter.getOrElse(TimeFormatter.instance)
      .getCollectionPeriod(size, getStart, getEnd, getDuration)

  override def getIntersection(other: ITimePeriod): ITimePeriod =
    Times.getIntersectionRange(this, other)

  override def getUnion(other: ITimePeriod): ITimePeriod =
    Times.getUnionRange(this, other)

  def indexOf(elem: ITimePeriod): Int =
    _periods.indexOf(elem)

  def removeAt(index: Int) {
    _periods.remove(index)
  }

  override def compareTo(other: ITimePeriod): Int = getStart compareTo other.getStart

  override def compare(other: ITimePeriod): Int = getStart compareTo other.getStart

  override def equals(other: Any): Boolean = {
    (other != null) &&
      (other.isInstanceOf[ITimePeriod]) &&
      isSampePeriod(other.asInstanceOf[ITimePeriod])
  }

  override def hashCode =
    ScalaHash.compute(getStart, getEnd, isReadonly)

  override def toString: String =
    getDescription()


}

